src/transaction.js

```diff
@@ -441,15 +438,10 @@ Transaction.prototype.toBuffer = function (buffer, initialOffset) {
     writeVarSlice(txOut.script)
   })
 
-  if (serializeWitnesses) {
-    var isNull = true
+  if (hasWitnesses) {
     this.ins.forEach(function (input) {
       writeVector(input.witness)
-      isNull = isNull && input.witness.length === 0
     })
-    if (isNull) {
-      throw new Error('Transaction has superfluous witness data')
-    }
```

Comment:
How does it know to call `.setWitness(i, EMPTY_WITNESS)` ? 

It's up to the caller to decide what way to serialize for witness. Separate issue, the `isNull` check only prevents a non-segwit transaction being serialized as a segwit one. The checks prevent an invalid case from being serialized/deserialized.